### PR TITLE
ref(ts) Convert SelectField to typescript

### PR DESCRIPTION
**Pull request by markstory**
I had to adjust the option format for incident rules as `Choices` doesn't include null and adding null spirals out to lots of other components.

### CODE REVIEW

File: src/sentry/static/sentry/app/views/settings/organizationDeveloperSettings/permissionSelection.tsx [hunk 1]
```diff
@@ -81,7 +81,7 @@ import SelectField from 'app/views/settings/components/forms/selectField';
 type Props = {
   permissions: Permissions;
   onChange: (permissions: Permissions) => void;
-  appPublished: Boolean;
```
Comment (markstory) [inline src/sentry/static/sentry/app/views/settings/organizationDeveloperSettings/permissionSelection.tsx:84]: This was caught as `unknown` is incompatible with `boolean` 😁

File: src/sentry/static/sentry/app/views/settings/components/forms/selectField.tsx [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+    /**
+     * Should the select allow empty values.
+     */
+    allowEmpty?: boolean;
+    small?: boolean;
+    /**
+     * A label that is shown inside the select control.
+     */
+    inFieldLabel?: string;
+    stacked?: boolean;
```
Comment (EvanPurkhiser) [inline src/sentry/static/sentry/app/views/settings/components/forms/selectField.tsx:25]: Doesn't this get pulled in from somewhere?

File: src/sentry/static/sentry/app/views/settings/components/forms/selectField.tsx [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+    onChange?.(value, {});
+    onBlur?.(value, {});
+  };
+
+  render() {
+    const {multiple, allowClear, ...otherProps} = this.props;
+    return (
+      <InputField
+        {...otherProps}
+        alignRight={this.props.small}
```
Comment (priscilawebdev) [inline src/sentry/static/sentry/app/views/settings/components/forms/selectField.tsx:87]: shouldn't we destructure `small` in the `this.props` above?

File: src/sentry/static/sentry/app/views/settings/components/forms/selectField.tsx [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+     * A label that is shown inside the select control.
+     */
+    inFieldLabel?: string;
+    stacked?: boolean;
+  };
+
+function getChoices<T>(props: Props<T>): Choices {
+  const choices = props.choices;
+  if (typeof choices === 'function') {
+    return choices(props);
```
Comment (priscilawebdev) [inline src/sentry/static/sentry/app/views/settings/components/forms/selectField.tsx:31]: should we omit `choices` in this `props`?

Comment (markstory) [reply src/sentry/static/sentry/app/views/settings/components/forms/selectField.tsx:31]: I didn't want to change the runtime code for this. I'd like to remove `choices` entirely but there is more work to do there.c

File: src/sentry/static/sentry/app/views/settings/components/forms/selectField.tsx (hunk 2 above)
Comment (markstory) [reply src/sentry/static/sentry/app/views/settings/components/forms/selectField.tsx:25]: It does. I'll take it out of this prop list.

**Review by evanpurkhiser — APPROVED** [review]

File: src/sentry/static/sentry/app/views/settings/components/forms/selectField.tsx [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+type InputFieldProps = React.ComponentProps<typeof InputField>;
+
+type Props<OptionType> = InputFieldProps &
+  Omit<ControlProps<OptionType>, 'onChange'> & {
+    /**
+     * Should the select be clearable?
+     */
+    allowClear?: boolean;
+    /**
+     * Should the select allow empty values.
```
Comment (priscilawebdev) [inline src/sentry/static/sentry/app/views/settings/components/forms/selectField.tsx:17]: should we add `?` here?
```suggestion
     * Should the select allow empty values?
```

File: src/sentry/static/sentry/app/views/settings/components/forms/selectField.tsx [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+     */
+    inFieldLabel?: string;
+  };
+
+function getChoices<T>(props: Props<T>): Choices {
+  const choices = props.choices;
+  if (typeof choices === 'function') {
+    return choices(props);
+  }
+  if (typeof choices === 'undefined') {
```
Comment (priscilawebdev) [inline src/sentry/static/sentry/app/views/settings/components/forms/selectField.tsx:32]:
```suggestion
  if (choices === undefined) {
```

**Review by priscilawebdev — APPROVED** [review]
left 2 suggestions but it looks good

**Pull request merged**